File: tree/java/leetcode94_Binary_Tree_Inorder_Traversal.java
//Given a binary tree, return the inorder traversal of its nodes' values.
//
//Example:
//
//Input: [1,null,2,3]
//   1
//    \
//     2
//    /
//   3
//
//Output: [1,3,2]

package wenxinjie1;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class leetcode94_Binary_Tree_Inorder_Traversal {
    public List<Integer> inorderTraversal(TreeNode root) {
    List<Integer> list = new ArrayList<Integer>();

    Stack<TreeNode> stack = new Stack<TreeNode>();
    TreeNode cur = root;

    while(cur!=null || !stack.empty()){
        while(cur!=null){
            stack.add(cur);
            cur = cur.left;
        }
        cur = stack.pop();
        list.add(cur.val);
        cur = cur.right;
    }

    return list;
    }
}

//Time: O(n)
//Space: O(1)
//Difficulty: medium

File: tree/python/leetcode94_Binary_Tree_Inorder_Traversal.py
# Given a binary tree, return the inorder traversal of its nodes' values.

# Example:

# Input: [1,null,2,3]
#    1
#     \
#      2
#     /
#    3

# Output: [1,3,2]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:

	# Solution 1: dfs
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None: return []
        res = []
        self.helper(res,root)
        return res
    
    def helper(self,res,root):
        if root is None: return
        self.helper(res,root.left)
        res.append(root.val)
        self.helper(res,root.right)


    # Solution 2: bfs
	def inorderTraversal2(self, root):
        stack = []
        cur = root
        res = []
        while stack or cur:
            while cur:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            res.append(cur.val)
            cur = cur.right
        return res


# Time: O(n)
# Space: O(1)
# Difficulty: medium
            